### PR TITLE
feat(controller): setup for .proto/grpc/http

### PR DESCRIPTION
**Pull request by charley-geoffroy**
## What does this PR do ?

This PR implements basic setup for orka controller :

- HTTP REST API
- gRPC  Server  for debug
- gRPC Client
- Proto buffers

This PR aims to get basic working setup on what could be controller part by making basic communication tests. 
You can also see the original POC respository [here](https://github.com/charley04310/poc-orka-controller-grpc)

### How should this be manually tested ?

You can refer to this [README](https://github.com/charley04310/poc-orka-controller-grpc/tree/main/controller)

### CODE REVIEW

**Comment by sameo** [comment]
Is that an implementation for https://github.com/dev-sys-do/orka/pull/7 ?

**Comment by charley-geoffroy** [comment]
Yes, it is the basic setup to start working on the #7

**Comment by sameo** [comment]
> Yes, it is the basic setup to start working on the #7

Good, thanks. The PR looks good, and please keep it up to date with #7 changes. We'll merge the former once the latter is merged.

**Comment by thomas-mauran** [comment]
> > Yes, it is the basic setup to start working on the #7
> 
> Good, thanks. The PR looks good, and please keep it up to date with #7 changes. We'll merge the former once the latter is merged.

Thanks, we wanted to merge this pr as soon as possible since it is only a basic grpc / http setup. This way we (the controller team) can all work on our forks with no git conflicts

File: controller/src/client.rs [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+
+    let request = SchedulingRequest {
+        workload: Some(workload),
+    };
+
+    let response = client.schedule(request).await?;
+
+    println!("RESPONSE={:?}", response);
+
+    Ok(())
```
Comment (sameo) [inline controller/src/client.rs]: Is this going to be a sample for a controller client?

File: controller/README.md [hunk 2]
```diff
@@ -0,0 +1,40 @@
+# Controller gRPC implementing task scheduling
+
+## Description
+
+This project is a gRPC implementation (server and client) that allows you to manage the state of a Kubernetes cluster. The gRPC server (controller) receives information about the current state of the cluster from the scheduler, saves this new state in a database, and retrieves the requested state from the database (workload). The controller then uses the scheduling function to schedule changes if the state is not good.
```
Comment (sameo) [inline controller/README.md]:
```suggestion
This is the `orka` controller reference implementation. The `orka` controller manages the overall state of an `orka` cluster through one public `gRPC` API and an internal one. The former allows for clients to control an `orka` cluster and e.g. create new workload instances, while the latter is consumed by the scheduler component to synchronously send node and instances status updates.

This project provides both a public API sample implementation and the internal API reference implementation.
```

File: controller/README.md [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+
+## Installation
+
+>**_Warning_**: When installing protocol buffers compiler ([protoc](https://grpc.io/docs/protoc-installation/)), make sure to install at least the version 3.15.x . Watch out by default `apt` installs version 3.12.x. To install the latest version required you can follow the steps bellow.
```
Comment (sameo) [inline controller/README.md]:
```suggestion
>**_Warning_**: The protocol buffers compiler ([protoc](https://grpc.io/docs/protoc-installation/)) version 3.15.x or later is required to build this project. Follow the steps below to install it manually if your distribution installs an earlier version of the `protoc` toolchain.
```

File: controller/src/server.rs [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+use axum::{response::Html, routing::get, Router};
+use std::net::SocketAddr;
+use tokio::task;
+
+pub mod scheduler {
+    tonic::include_proto!("orkascheduler"); // The string specified here must match the proto package name
+}
+
+#[derive(Debug, Default)]
+pub struct MySchedulingService {}
```
Comment (sameo) [inline controller/src/server.rs]: Could you rename that structure to something more generic like e.g. `SchedulerService` ?

File: controller/src/server.rs [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+            rejection_reason: Some(1),
+        };
+
+        Ok(Response::new(response))
+    }
+}
+
+#[tokio::main]
+async fn main() -> Result<(), Box<dyn std::error::Error>> {
+    // GRPC
```
Comment (sameo) [inline controller/src/server.rs]: That's not a really useful comment, imo. Something like `Initialize the gRPC server` would be better.

File: controller/src/server.rs [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+    let grpc_addr = "[::1]:50051".parse()?;
+    let scheduler = MySchedulingService::default();
+
+    // Spawn the gRPC server as a tokio task
+    let grpc_thread = task::spawn(async move {
+        println!("Running grpc here: {}", grpc_addr);
```
Comment (sameo) [inline controller/src/server.rs]: I'm ok with `println` for now, but you should use a more flexible and robust logging mechanism. See e.g. https://docs.rs/log/latest/log/

File: controller/src/server.rs [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+        Server::builder()
+            .add_service(SchedulingServiceServer::new(scheduler))
+            .serve(grpc_addr)
+            .await
+            .unwrap();
+    });
+
+    // HTTP
```
Comment (sameo) [inline controller/src/server.rs]: Same comment as before.

File: controller/src/client.rs (hunk 1 above)
Comment (thomas-mauran) [reply controller/src/client.rs]: Yes, this is a basic example of grpc client for the controller, the goal is to communicate with grpc to the scheduler. In the next version of the code we are splitting the code in different files and we will try to make rust test for each method

File: controller/README.md (hunk 2 above)
Comment (charley-geoffroy) [reply controller/README.md]: Done :+1:

File: controller/README.md (hunk 3 above)
Comment (charley-geoffroy) [reply controller/README.md]: Done :+1:

File: controller/src/server.rs (hunk 4 above)
Comment (charley-geoffroy) [reply controller/src/server.rs]: Done :+1:

File: controller/src/server.rs (hunk 5 above)
Comment (charley-geoffroy) [reply controller/src/server.rs]: Done :+1:

File: controller/src/server.rs (hunk 6 above)
Comment (charley-geoffroy) [reply controller/src/server.rs]: You're absolutely correct! We've introduced the pretty_env_logger = "0.5.0" dependency to ensure a more logical logging approach for the various controller messages. :+1: Moreover, we've updated the .README file to provide the necessary configuration instructions.

File: controller/src/server.rs (hunk 7 above)
Comment (charley-geoffroy) [reply controller/src/server.rs]: Done :+1:

**Review by sameo — APPROVED** [review]

**Pull request merged**